File: ml_backend/database_service.py
import psycopg2
import psycopg2.extras
import os
from datetime import datetime, timedelta
import json

class DatabaseService:
    def __init__(self):
        self.connection = None
        self.connect()
    
    def connect(self):
        """Connect to PostgreSQL database"""
        try:
            database_url = os.getenv('DATABASE_URL', 'postgresql://postgres:password@localhost:5432/mental_health_companion')
            self.connection = psycopg2.connect(database_url)
            print("✅ Connected to PostgreSQL database")
        except Exception as e:
            print(f"❌ Error connecting to database: {e}")
            self.connection = None
    
    def get_user_mood_entries(self, user_id, limit=100):
        """Get mood entries for a specific user"""
        if not self.connection:
            return []
        
        try:
            with self.connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cursor:
                cursor.execute("""
                    SELECT mood, emotions, note, date, created_at
                    FROM "MoodEntry"
                    WHERE "userId" = %s
                    ORDER BY date DESC
                    LIMIT %s
                """, (user_id, limit))
                return cursor.fetchall()
        except Exception as e:
            print(f"Error fetching mood entries: {e}")
            return []
    
    def get_user_journal_entries(self, user_id, limit=100):
        """Get journal entries for a specific user"""
        if not self.connection:
            return []
        
        try:
            with self.connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cursor:
                cursor.execute("""
                    SELECT title, content, date, created_at
                    FROM "JournalEntry"
                    WHERE "userId" = %s
                    ORDER BY date DESC
                    LIMIT %s
                """, (user_id, limit))
                return cursor.fetchall()
        except Exception as e:
            print(f"Error fetching journal entries: {e}")
            return []
    
    def get_user_chat_messages(self, user_id, limit=100):
        """Get chat messages for a specific user"""
        if not self.connection:
            return []
        
        try:
            with self.connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cursor:
                cursor.execute("""
                    SELECT message, "isUser", timestamp, created_at
                    FROM "ChatMessage"
                    WHERE "userId" = %s
                    ORDER BY timestamp DESC
                    LIMIT %s
                """, (user_id, limit))
                return cursor.fetchall()
        except Exception as e:
            print(f"Error fetching chat messages: {e}")
            return []
    
    def get_user_data_for_ml(self, user_id):
        """Get all user data formatted for ML processing"""
        mood_entries = self.get_user_mood_entries(user_id)
        journal_entries = self.get_user_journal_entries(user_id)
        chat_messages = self.get_user_chat_messages(user_id)
        
        return {
            'mood_entries': mood_entries,
            'journal_entries': journal_entries,
            'chat_messages': chat_messages
        }
    
    def close(self):
        """Close database connection"""
        if self.connection:
            self.connection.close()
            print("Database connection closed")

File: lib/database-service.ts
import { prisma } from './db'
// import { getServerSession } from 'next-auth'
// import { authOptions } from './auth'

// Mood Entry Services
export async function createMoodEntry(data: {
  mood: number
  emotions: string[]
  note?: string
  userId: string
}) {
  return await prisma.moodEntry.create({
    data: {
      mood: data.mood,
      emotions: data.emotions,
      note: data.note ?? null,
      userId: data.userId,
    },
  })
}

export async function getMoodEntries(userId: string, limit?: number) {
  return await prisma.moodEntry.findMany({
    where: { userId },
    orderBy: { date: 'desc' },
    ...(limit && { take: limit }),
  })
}

export async function getMoodEntriesByDateRange(userId: string, startDate: Date, endDate: Date) {
  return await prisma.moodEntry.findMany({
    where: {
      userId,
      date: {
        gte: startDate,
        lte: endDate,
      },
    },
    orderBy: { date: 'desc' },
  })
}

// Journal Entry Services
export async function createJournalEntry(data: {
  title?: string
  content: string
  userId: string
}) {
  return await prisma.journalEntry.create({
    data: {
      title: data.title ?? null,
      content: data.content,
      userId: data.userId,
    },
  })
}

export async function getJournalEntries(userId: string, limit?: number) {
  return await prisma.journalEntry.findMany({
    where: { userId },
    orderBy: { date: 'desc' },
    ...(limit && { take: limit }),
  })
}

export async function getJournalEntryById(id: string, userId: string) {
  return await prisma.journalEntry.findFirst({
    where: { id, userId },
  })
}

export async function updateJournalEntry(id: string, userId: string, data: {
  title?: string
  content: string
}) {
  return await prisma.journalEntry.updateMany({
    where: { id, userId },
    data,
  })
}

export async function deleteJournalEntry(id: string, userId: string) {
  return await prisma.journalEntry.deleteMany({
    where: { id, userId },
  })
}

// Chat Message Services
export async function createChatMessage(data: {
  message: string
  isUser: boolean
  userId: string
}) {
  return await prisma.chatMessage.create({
    data: {
      message: data.message,
      isUser: data.isUser,
      userId: data.userId,
    },
  })
}

export async function getChatMessages(userId: string, limit?: number) {
  return await prisma.chatMessage.findMany({
    where: { userId },
    orderBy: { timestamp: 'desc' },
    ...(limit && { take: limit }),
  })
}

export async function getChatMessagesByDateRange(userId: string, startDate: Date, endDate: Date) {
  return await prisma.chatMessage.findMany({
    where: {
      userId,
      timestamp: {
        gte: startDate,
        lte: endDate,
      },
    },
    orderBy: { timestamp: 'desc' },
  })
}

// User Services
export async function getUserById(id: string) {
  return await prisma.user.findUnique({
    where: { id },
    select: {
      id: true,
      name: true,
      email: true,
      createdAt: true,
    },
  })
}

export async function getUserStats(userId: string) {
  const [moodCount, journalCount, chatCount] = await Promise.all([
    prisma.moodEntry.count({ where: { userId } }),
    prisma.journalEntry.count({ where: { userId } }),
    prisma.chatMessage.count({ where: { userId } }),
  ])

  return {
    moodEntries: moodCount,
    journalEntries: journalCount,
    chatMessages: chatCount,
  }
}
